File: backend/src/iaas_sim/application/operation.py
from __future__ import annotations

from dataclasses import dataclass
from typing import NewType, Protocol

from iaas_sim.application.identity import BackendVirtualMachineRef
from iaas_sim.domain.entity.operation import Operation, OperationId
from iaas_sim.domain.entity.virtual_machine import VirtualMachineId
from iaas_sim.result import Result

BackendOperationRef = NewType("BackendOperationRef", str)


@dataclass(frozen=True, slots=True)
class BackendOperationRunning:
    pass


@dataclass(frozen=True, slots=True)
class BackendOperationNoResult:
    pass


@dataclass(frozen=True, slots=True)
class BackendVirtualMachineCreated:
    backend_ref: BackendVirtualMachineRef


type BackendOperationResult = BackendOperationNoResult | BackendVirtualMachineCreated


@dataclass(frozen=True, slots=True)
class BackendOperationSucceeded:
    result: BackendOperationResult = BackendOperationNoResult()


@dataclass(frozen=True, slots=True)
class BackendOperationFailed:
    reason: str


type BackendOperationStatus = (
    BackendOperationRunning | BackendOperationSucceeded | BackendOperationFailed
)


@dataclass(frozen=True, slots=True)
class StoredOperation:
    operation: Operation
    backend_ref: BackendOperationRef


@dataclass(frozen=True, slots=True)
class OperationNotFound:
    operation_id: OperationId


@dataclass(frozen=True, slots=True)
class OperationPersistenceFailure:
    operation: str
    reason: str


@dataclass(frozen=True, slots=True)
class OperationPollingFailure:
    reason: str


type OperationStoreError = OperationNotFound | OperationPersistenceFailure


class OperationStorePort(Protocol):
    def create_running(
        self, operation: Operation, backend_ref: BackendOperationRef
    ) -> Result[Operation, OperationPersistenceFailure]: ...
    def get(self, operation_id: OperationId) -> Result[StoredOperation, OperationStoreError]: ...
    def complete(self, operation: Operation) -> Result[Operation, OperationStoreError]: ...


class VirtualMachineCreateFinalizerPort(Protocol):
    def finalize_virtual_machine_create(
        self,
        operation: Operation,
        virtual_machine_id: VirtualMachineId,
        backend_ref: BackendVirtualMachineRef,
    ) -> Result[Operation, OperationPersistenceFailure]: ...


class BackendOperationPort(Protocol):
    def get_operation_status(
        self, backend_ref: BackendOperationRef
    ) -> Result[BackendOperationStatus, OperationPollingFailure]: ...

File: backend-kotlin/src/main/kotlin/dev/iaassim/application/Operation.kt
package dev.iaassim.application

import dev.iaassim.domain.entity.operation.Operation
import dev.iaassim.domain.entity.operation.OperationId
import dev.iaassim.result.Outcome

@JvmInline value class BackendOperationRef(val value: String)
sealed interface BackendOperationStatus
data object BackendOperationRunning : BackendOperationStatus
data object BackendOperationSucceeded : BackendOperationStatus
data class BackendOperationFailed(val reason: String) : BackendOperationStatus

data class OperationNotFound(val operationId: OperationId) : OperationStoreError, GetOperationError
data class OperationPersistenceFailure(val operation: String, val reason: String) : OperationStoreError, GetOperationError
data class OperationPollingFailure(val reason: String) : GetOperationError
sealed interface OperationStoreError
sealed interface GetOperationError
data class StoredOperation(val operation: Operation, val backendRef: BackendOperationRef)

interface OperationStorePort {
    fun createRunning(operation: Operation, backendRef: BackendOperationRef): Outcome<Operation, OperationPersistenceFailure>
    fun get(operationId: OperationId): Outcome<StoredOperation, OperationStoreError>
    fun complete(operation: Operation): Outcome<Operation, OperationStoreError>
}

interface BackendOperationPort {
    fun getOperationStatus(backendRef: BackendOperationRef): Outcome<BackendOperationStatus, OperationPollingFailure>
}
